lib/winappdriver.js

```diff
@@ -1,36 +1,250 @@
-﻿import { JWProxy } from 'appium-base-driver';
-import logger from './logger';
+﻿import events from 'events';
+import { JWProxy } from 'appium-base-driver';
+import log from './logger';
+import { SubProcess } from 'teen_process';
+import { getWinAppDriverBinaryPath } from './install';
+import { retry, retryInterval } from 'asyncbox';
+import { system, fs } from 'appium-support';
 
-const REQD_PARAMS = ['app', 'host', 'port'];
 
-class WinAppDriverServer {
+
+const REQD_PARAMS = ['app'];
+const DEFAULT_HOST = '127.0.0.1';
+const DEFAULT_PORT = "4823"; //  should be non-4723 to avoid conflict on the same box
+
+class WinAppDriver extends events.EventEmitter {
   constructor(opts = {}) {
+    const {proxyhost, proxyport, executable, cmdArgs} = opts;
+    super();
+
     for (let req of REQD_PARAMS) {
       if (!opts || !opts[req]) {
         throw new Error(`Option '${req}' is required!`);
       }
       this[req] = opts[req];
     }
 
-    this.jwproxy = new JWProxy({ server: this.host, port: this.port });
-    this.proxyReqRes = this.jwproxy.proxyReqRes.bind(this.jwproxy);
+    this.proxyHost = proxyhost || DEFAULT_HOST;
+    this.proxyPort = proxyport || DEFAULT_PORT;
+    this.cmdArgs = cmdArgs;
+    this.winappdriver = executable;
+    this.proc = null;
+    this.executableVerified = false;
+    this.state = WinAppDriver.STATE_STOPPED;
+    this.jwproxy = new JWProxy({server: this.proxyHost, port: this.proxyPort});
+}
+
+  async startWinAppDriverProcess (caps) {
+    this.initWinAppDriverPath();
+    return this.start(caps, true);
+  }
+
+  async initWinAppDriverPath () {
+    if (this.executableVerified) return;
+    
+    let binPath = this.winappdriver || (await getWinAppDriverBinaryPath());
+    if (!await fs.exists(binPath)) {
+      throw new Error(`Trying to use a WinAppDriver binary at the path ` +
+                      `${binPath}, but it doesn't exist!`);
+    }
+    this.winappdriverpath = binPath;
+    this.executableVerified = true;
+    log.info(`Set WinAppDriver binary as: ${this.winappdriverpath}`);
+  }
+
+  async start (caps, emitStartingState = true) {
+
+    this.capabilities = caps;
+    if (emitStartingState) {
+      this.changeState(WinAppDriver.STATE_STARTING);
+    }
+
+    // XXXYD TODO: would be better if WinAppDriver didn't require passing in /wd/hub as a param
+    let args = [this.proxyPort + "/wd/hub"];
+
+    if (this.cmdArgs) {
+      args = args.concat(this.cmdArgs);
+    }
+
+    const startDetector = (stdout) => {
+        // XXXYD What is this for, looking for a log entry that shows we're up?
+        // return stdout.indexOf('ENTER') != -1;
```

Comment:
it was the UCS2 text that caused indexOf to fail, updating PR with fix, and we'll need to come back and fix again once WAD starts outputting utf-8